applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_residual_criteria.h

```diff
@@ -73,81 +124,198 @@ class TrilinosResidualCriteria : public ResidualCriteria< TSparseSpace, TDenseSp
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
     }
 
     ///@}
+    ///@name Friends
+    ///@{
 
-private:
+    ///@}
+protected:
+    ///@name Protected Operations
+    ///@{
 
-    ///@name Member Variables
+    /**
+     * @brief This method computes the active dofs
+     * @param rModelPart Reference to the ModelPart containing the problem
+     * @param rDofSet The whole set of dofs
+     */
+    void ComputeActiveDofs(
+        ModelPart& rModelPart,
+        const ModelPart::DofsArrayType& rDofSet
+        ) override
+    {
+        // Filling mActiveDofs when MPC exist
+        ConstraintUtilities::DistributedComputeActiveDofs(rModelPart, BaseType::mActiveDofs, rDofSet, mInitialDoFId);
+    }
+
+    /**
+     * @brief Check if a Degree of Freedom (Dof) is active
+     * @details This function checks if a given Degree of Freedom (Dof) is active.
+     * @param rDof The Degree of Freedom to check.
+     * @param Rank The rank of the Dof.
+     * @return True if the Dof is free, false otherwise.
+     */
+    bool IsActiveDof(
+        const DofType& rDof,
+        const int Rank
+        ) override
+    {
+        const IndexType dof_id = rDof.EquationId();
+        KRATOS_DEBUG_ERROR_IF((dof_id - mInitialDoFId) >= BaseType::mActiveDofs.size()) << "DofId is greater than the size of the active Dofs vector. DofId: " << dof_id << "\tInitialDoFId: " << mInitialDoFId << "\tActiveDofs size: " << BaseType::mActiveDofs.size() << std::endl;
+        return (BaseType::mActiveDofs[dof_id - mInitialDoFId] == 1 && (rDof.GetSolutionStepValue(PARTITION_INDEX) == Rank));
+    }
+
+    /**
+     * @brief Check if a Degree of Freedom (Dof) is free.
+     * @details This function checks if a given Degree of Freedom (Dof) is free.
+     * @param rDof The Degree of Freedom to check.
+     * @param Rank The rank of the Dof.
+     * @return True if the Dof is free, false otherwise.
+     */
+    bool IsFreeDof(
+        const DofType& rDof,
+        const int Rank
+        ) override
+    {
+        return (rDof.IsFree() && (rDof.GetSolutionStepValue(PARTITION_INDEX) == Rank));
+    }
+
+    /**
+     * @brief This method assigns settings to member variables
+     * @param ThisParameters Parameters that are assigned to the member variables
+     */
+    void AssignSettings(const Parameters ThisParameters) override
+    {
+        BaseType::AssignSettings(ThisParameters);
+    }
+
+    ///@}
+    ///@name Protected  Access
     ///@{
 
     ///@}
-    ///@name Private Operations
+    ///@name Protected Inquiry
     ///@{
 
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+private:
+    ///@name Member Variables
+    ///@{
+
+    IndexType mInitialDoFId = 0; /// The initial DoF Id
```

Comment:
i would start this with an "impossible" is, say -1

Reply:
-1 is not possible for unsigned int (IndexType -> std::size_t)

Reply:
Okay, I can use the lasrgest number then